Document project structure in README.md

kirschmet2 opened the pull request:
- explain purpose of individual files

kirschmet2 committed:
Document project structure in README.md

- explain purpose of individual files

kirschmet2 committed:
Mention new 'Adders' module in documentation

ners commented:
This is very well written, thank you!

Some style tips to consider:
- we could use Markdown's inline code style (with \`backticks\`) to denote directories and files
- we could link the directories and files

Instead of having per-file documentation in the readme, we should rather have those in the Hs files themselves at some point. The most important questions we want to answer are for example "where do I start looking if I want to know how this works?" "where do I add a new logic function?" etc. On the other hand, the Nix files are described very well. Perhaps we should simply list the files and directories an the root of the project, and dedicate a paragraph to each.

kirschmet2 committed:
Remove reference to individual source files, add links

kirschmet2 committed:
Attempt to correct link syntax

kirschmet2 committed:
Fixed links for real this time - sorry!

ners commented on README.md:
```diff
@@ -16,6 +16,17 @@ A possible outcome would be an emulator that can run RV32I assembly.
+  - [`src/`](/src): Haskell library folder, each file inside is a Haskell module providing functions-
+  - [`test/`](/test) : Contains Haskell source for an executable that is built and run by `cabal test`, used to test the functions defined in `src/`.
+  - [`CHANGELOG.md`](CHANGELOG.md): This file will only become relevant in case of a release.
+  - [`flake.nix`](flake.nix): This is a [Nix flake](https://nixos.wiki/wiki/Flakes) providing dependencies for the project, namely GHC and Cabal, for Linux and MacOS/Darwin on x86_64.
```
I wouldn't bake the supported platforms into the readme. At the moment we specify those two, but in the future we'll probably also add support for ARM and, of course, RISC-V. :)

kirschmet2 replied:
Ok, I guess we can just remove it or add `for now` at the end. Are there conventions on when/how often to update a readme?

ners commented on README.md:
```diff
@@ -16,6 +16,17 @@ A possible outcome would be an emulator that can run RV32I assembly.
+  - [`test/`](/test) : Contains Haskell source for an executable that is built and run by `cabal test`, used to test the functions defined in `src/`.
+  - [`CHANGELOG.md`](CHANGELOG.md): This file will only become relevant in case of a release.
+  - [`flake.nix`](flake.nix): This is a [Nix flake](https://nixos.wiki/wiki/Flakes) providing dependencies for the project, namely GHC and Cabal, for Linux and MacOS/Darwin on x86_64.
+  - [`flake.lock`](flake.lock): File automatically created by Nix that does not need to be touched.
```
While it's true this file does not need to be touched, it'd be nice to say what purpose it serves, and link to some documentation, just like you have for Cabal.

kirschmet2 replied:
Like this?
```Markdown
  - [`flake.lock`](flake.lock): File specifying the exact versions of dependencies, created by the [`nix flake lock`](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-flake-lock.html) command.
```

ners commented on README.md:
```diff
@@ -16,6 +16,17 @@ A possible outcome would be an emulator that can run RV32I assembly.
+  - [`flake.lock`](flake.lock): File automatically created by Nix that does not need to be touched.
+  - [`LICENSE`](LICENSE): The Apache 2.0 license.
+  - [`riscvs-hs.cabal`](riscvs-hs.cabal): Cabal [package description](https://cabal.readthedocs.io/en/3.4/cabal-package.html).
+  - [`shell.nix`](shell.nix): For compatibility with non-flake-enabled Nix versions.
```
Might be worth mentioning an example of this: VSCode's nix-environment-selector plugin. Lots of people use vscode, so it'd be nice to help them connect it with Nix.

kirschmet2 replied:
Thanks for the idea.
```Markdown
  - [`shell.nix`](shell.nix): For compatibility with non-flake-enabled Nix versions and tools requiring .nix files, such as the [Nix environment selector](https://github.com/arrterian/nix-env-selector) plugin for VSCode.
```

kirschmet2 committed:
Add links to documentation
- mention use case for shell.nix
- link to documentation of flake.lock
- remove list of supported platforms

fricklerhandwerk commented on README.md:
```diff
@@ -16,6 +16,18 @@ A possible outcome would be an emulator that can run RV32I assembly.
 
+# Project structure
+- The structure of the project is dictated by the tools we are using, namely Nix and Cabal. This repository is both a Nix flake and a Cabal package and as such contains the following folders/files:
+  - [`app/`](/app): Contains Haskell source for an executable that can be built and run by `cabal run`. It does not do anything meaningful at the moment.
```
```suggestion
  - [`app/`](/app): Haskell source code for an executable that can be built and run by `cabal run`.
```

It does not do anything right now, but we'd have to update it once it does, and I'll bet we'd forget that.

fricklerhandwerk commented on README.md:
```diff
@@ -16,6 +16,18 @@ A possible outcome would be an emulator that can run RV32I assembly.
+# Project structure
+- The structure of the project is dictated by the tools we are using, namely Nix and Cabal. This repository is both a Nix flake and a Cabal package and as such contains the following folders/files:
+  - [`app/`](/app): Contains Haskell source for an executable that can be built and run by `cabal run`. It does not do anything meaningful at the moment.
+  - [`src/`](/src): Haskell library folder, each file inside is a Haskell module providing functions-
```
```suggestion
  - [`src/`](/src): Haskell library folder, each file inside is a Haskell module.
```

fricklerhandwerk commented on README.md:
```diff
@@ -16,6 +16,18 @@ A possible outcome would be an emulator that can run RV32I assembly.
+- The structure of the project is dictated by the tools we are using, namely Nix and Cabal. This repository is both a Nix flake and a Cabal package and as such contains the following folders/files:
+  - [`app/`](/app): Contains Haskell source for an executable that can be built and run by `cabal run`. It does not do anything meaningful at the moment.
+  - [`src/`](/src): Haskell library folder, each file inside is a Haskell module providing functions-
+  - [`test/`](/test) : Contains Haskell source for an executable that is built and run by `cabal test`, used to test the functions defined in `src/`.
```
```suggestion
  - [`test/`](/test): Haskell source code for an executable that is built and run by `cabal test`, used to test the functions defined in `src/`.
```

fricklerhandwerk commented on README.md:
```diff
@@ -16,6 +16,18 @@ A possible outcome would be an emulator that can run RV32I assembly.
+  - [`app/`](/app): Contains Haskell source for an executable that can be built and run by `cabal run`. It does not do anything meaningful at the moment.
+  - [`src/`](/src): Haskell library folder, each file inside is a Haskell module providing functions-
+  - [`test/`](/test) : Contains Haskell source for an executable that is built and run by `cabal test`, used to test the functions defined in `src/`.
+  - [`CHANGELOG.md`](CHANGELOG.md): This file will only become relevant in case of a release.
```
Since it's rarely relevant anyway, we could move it to the end. I prefer sorting by relevance, but I agree that it may be harder to understand because the directory listing is usually alphabetical, so I don't insist on changing the order.

fricklerhandwerk commented on README.md:
```diff
@@ -16,6 +16,18 @@ A possible outcome would be an emulator that can run RV32I assembly.
+  - [`src/`](/src): Haskell library folder, each file inside is a Haskell module providing functions-
+  - [`test/`](/test) : Contains Haskell source for an executable that is built and run by `cabal test`, used to test the functions defined in `src/`.
+  - [`CHANGELOG.md`](CHANGELOG.md): This file will only become relevant in case of a release.
+  - [`flake.nix`](flake.nix): This is a [Nix flake](https://nixos.wiki/wiki/Flakes) providing dependencies for the project, namely GHC and Cabal.
```
```suggestion
  - [`flake.nix`](flake.nix): This is a [Nix flake](https://nixos.wiki/wiki/Flakes) providing dependencies for the project.
```

The details may change, and there is no point maintaining them in prose.

fricklerhandwerk commented on README.md:
```diff
@@ -16,6 +16,18 @@ A possible outcome would be an emulator that can run RV32I assembly.
+  - [`flake.lock`](flake.lock): File specifying the exact versions of dependencies, created by the [`nix flake lock`](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-flake-lock.html) command. 
+  - [`LICENSE`](LICENSE): The Apache 2.0 license.
+  - [`riscvs-hs.cabal`](riscvs-hs.cabal): Cabal [package description](https://cabal.readthedocs.io/en/3.4/cabal-package.html).
+  - [`shell.nix`](shell.nix): For compatibility with non-flake-enabled Nix versions and tools requiring .nix files, such as the [Nix environment selector](https://github.com/arrterian/nix-env-selector) plugin for VSCode.
```
```suggestion
  - [`shell.nix`](shell.nix): For compatibility with Nix without flakes, as well as other tools.
```

The details may change. Don't think "non-flake-enabled" is an actual word.

ners replied:
Let's explain how to connect VSCode with the Nix shell in a separate section, perhaps "Notes", "FAQ", or "Tips and tricks".

fricklerhandwerk reviewed (approved):
I agree with @ners, this is really well-written and precise. Only a few details on language and removing information that may change, because we don't want to keep that up to date manually.

kirschmet2 committed:
Remove reference to current situation subject to change

Co-authored-by: Valentin Gagarin <valentin@fricklerhandwerk.de>

kirschmet2 committed:
Remove unneccessary words

Co-authored-by: Valentin Gagarin <valentin@fricklerhandwerk.de>

kirschmet2 committed:
clarify wording

Co-authored-by: Valentin Gagarin <valentin@fricklerhandwerk.de>

kirschmet2 committed:
Remove reference to current situation subject to change

Co-authored-by: Valentin Gagarin <valentin@fricklerhandwerk.de>

kirschmet2 committed:
Create "Notes" section, move reference to VSCode plugin there
